### PR TITLE
Telescope display ret addr #1071

### LINKED ISSUE #1071: telescope: display ret address
It would be great to display return addresses in the `telescope`, `stack` etc outputs.

I initially worked on it in https://github.com/pwndbg/pwndbg/pull/794 but the patch had some issues.

We should proceed with:
1. Move the logic of finding return addresses to `pwndbg.stack.yield_return_addresses`, e.g.:
```py
def yield_return_addresses():
    sp = pwndbg.regs.sp
    stack = pwndbg.vmmap.find(sp)

    # Enumerate all return addresses
    frame = gdb.newest_frame()
    addresses = []
    while frame:
        addresses.append(frame.pc())
        frame = frame.older()

    # Find all of them on the stack
    start = stack.vaddr
    stop = start + stack.memsz
    while addresses and start < sp < stop:
        value = pwndbg.memory.u(sp)

        if value in addresses:
            index = addresses.index(value)
            del addresses[:index]
            yield sp

        sp += pwndbg.arch.ptrsize
```
* Question: shouldn't the first operator (between `start` and `sp`) here: `start < sp < stop` be `<=`? On a quick glance now I think so

2. Rethink how to add it into `telescope` and other output
* Shall we display it in all cases or only on demand?

3. Rethink if we want to add GDB variables for return addresses. This is really neat, but if we do it, we probably need to do it on each GDB stop and then it may be hurt performance on tough targets when debugging via gdbserver...

### PR DESCRIPTION
**Pull request by dmur1**
seeks to fix https://github.com/pwndbg/pwndbg/issues/1071

image

### CODE REVIEW

File: pwndbg/gdblib/stack.py [hunk 1]
```diff
@@ -135,3 +136,28 @@ def is_executable() -> bool:
...
 
     return not nx
+
+
+def yield_return_addresses():
+    sp = pwndbg.gdblib.regs.sp
+    stack = pwndbg.gdblib.vmmap.find(sp)
+
+    # Enumerate all return addresses
+    frame = gdb.newest_frame()
```
Comment (disconnect3d) [inline pwndbg/gdblib/stack.py:167]: What happens if GDB can't figure out stack frames?

File: pwndbg/gdblib/stack.py [hunk 2]
```diff
@@ -135,3 +136,28 @@ def is_executable() -> bool:
...
+    addresses = []
+    while frame:
+        addresses.append(frame.pc())
+        frame = frame.older()
+
+    # Find all of them on the stack
+    start = stack.vaddr
+    stop = start + stack.memsz
```
Comment (disconnect3d) [inline pwndbg/gdblib/stack.py]: Just `start, stop = stack.start, stack.end` I believe

**Comment by disconnect3d** [comment]
There are couple of things here:
1. Is this relevant no matter where return addresses are located? If we telescope a heap memory, do we want a return address to show there?
2. If we show return addresses, should we also show saved BP addresses if they are there? I am not sure but I bet the frame object has this info
3. We need to document this behavior, and likely allow the user to disable it
4. This will probably crash if the stack frame makes no sense for GDB. You should be able to try it out by doing `set $rsp=$rsp+8` or `set $rsp=0x1234` and seeing what it will do then

**Comment by disconnect3d** [comment]
@dmur1 ping :P

**Comment by OBarronCS** [comment]
This would be great to add. Perhaps a compromise for the issue of performance (reading the entire stack every time) would be to pass in a range to `yield_return_addresses`, like `yield_return_addresses(start,stop)`, so that when this is used in telescope or the context output, it will only ever search a very small range of addresses in most cases, and not the entire stack. Additionally, keeping it configurable at a global level, as well as perhaps a flag could exist to pass to telescope to disable it for a given run, too.

**Comment by disconnect3d** [comment]
@dmur1 any chance u can update this PR/fix conflicts etc?

**Comment by dmur1** [comment]
> @dmur1 any chance u can update this PR/fix conflicts etc?

yes!

File: pwndbg/gdblib/stack.py (hunk 1 above)
Comment (dmur1) [reply pwndbg/gdblib/stack.py:167]: not sure :man_shrugging: this pattern is used in a lot of places ( just getting the frame and assuming its valid ) fwiw

File: pwndbg/gdblib/stack.py (hunk 2 above)
Comment (dmur1) [reply pwndbg/gdblib/stack.py]: https://github.com/pwndbg/pwndbg/pull/1889/commits/5d5eac9369b9b0f5424170f3d98661c86ba00c70 resolved

**Comment by disconnect3d** [comment]
@OBarronCS would u perhaps want to make a review of this PR? :)

**Comment by disconnect3d** [comment]
@dmur1 RE: non-existent stack frame - run `gdb --quiet /bin/ls` and then:

```
pwndbg> set context-sections ''
...
pwndbg> entry
...
pwndbg> stack
00:0000│ r13 rsp ret0 0x7fffffffe370 ◂— 0x1
01:0008│ ret1         0x7fffffffe378 —▸ 0x7fffffffe604 ◂— '/usr/bin/ls'
02:0010│ ret2         0x7fffffffe380 ◂— 0x0
03:0018│ rcx          0x7fffffffe388 —▸ 0x7fffffffe610 ◂— 'SHELL=/bin/bash'
04:0020│              0x7fffffffe390 —▸ 0x7fffffffe620 ◂— 'PWD=/root/pwndbg'
05:0028│              0x7fffffffe398 —▸ 0x7fffffffe631 ◂— 'NIX_PROFILES=/nix/var/nix/profiles/default /root/.nix-profile'
06:0030│              0x7fffffffe3a0 —▸ 0x7fffffffe66f ◂— 'LOGNAME=root'
07:0038│              0x7fffffffe3a8 —▸ 0x7fffffffe67c ◂— 'XDG_SESSION_TYPE=tty'

pwndbg> set $rsp=0x1234

pwndbg> stack
Exception occurred: telescope: 'NoneType' object has no attribute 'start' (<class 'AttributeError'>)
For more info invoke `set exception-verbose on` and rerun the command
or debug it by yourself with `set exception-debugger on`
pwndbg> set exception-verbose on
Set whether to print a full stacktrace for exceptions raised in Pwndbg commands to 'on'.
pwndbg> stack
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /root/pwndbg/pwndbg/commands/__init__.py:190 in __call__                                         │
│                                                                                                  │
│   187 │                                                                                          │
│   188 │   def __call__(self, *args: Any, **kwargs: Any) -> str | None:                           │
│   189 │   │   try:                                                                               │
│ ❱ 190 │   │   │   return self.function(*args, **kwargs)                                          │
│   191 │   │   except TypeError as te:                                                            │
│   192 │   │   │   print(f"{self.function.__name__.strip()!r}: {self.function.__doc__.strip()}"   │
│   193 │   │   │   pwndbg.exception.handle(self.function.__name__)                                │
│                                                                                                  │
│ /root/pwndbg/pwndbg/commands/__init__.py:346 in _OnlyWhenRunning                                 │
│                                                                                                  │
│   343 │   @functools.wraps(function)                                                             │
│   344 │   def _OnlyWhenRunning(*a: Any, **kw: Any) -> Optional[T]:                               │
│   345 │   │   if pwndbg.gdblib.proc.alive:                                                       │
│ ❱ 346 │   │   │   return function(*a, **kw)                                                      │
│   347 │   │   else:                                                                              │
│   348 │   │   │   print(f"{function.__name__}: The program is not being run.")                   │
│   349 │   │   │   return None                                                                    │
│                                                                                                  │
│ /root/pwndbg/pwndbg/commands/telescope.py:159 in telescope                                       │
│                                                                                                  │
│   156 │   for reg in pwndbg.gdblib.regs.common:                                                  │
│   157 │   │   name_values[pwndbg.gdblib.regs[reg]].append(reg)                                   │
│   158 │                                                                                          │
│ ❱ 159 │   for idx, retaddr in enumerate(pwndbg.gdblib.stack.yield_return_addresses()):           │
│   160 │   │   name_values[retaddr].append(f"ret{idx}")                                           │
│   161 │                                                                                          │
│   162 │   if not inverse:                                                                        │
│                                                                                                  │
│ /root/pwndbg/pwndbg/gdblib/stack.py:171 in yield_return_addresses                                │
│                                                                                                  │
│   168 │   │   frame = frame.older()                                                              │
│   169 │                                                                                          │
│   170 │   # Find all of them on the stack                                                        │
│ ❱ 171 │   start, stop = stack.start, stack.end                                                   │
│   172 │   while addresses and start < sp < stop:                                                 │
│   173 │   │   value = pwndbg.gdblib.memory.u(sp)                                                 │
│   174                                                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'NoneType' object has no attribute 'start'
```

Pls fix :)

File: pwndbg/commands/telescope.py [hunk 3]
```diff
@@ -153,10 +152,12 @@ def telescope(
         count -= address
         count = max(math.ceil(count / ptrsize), 1)
 
-    # Map of address to register string
-    reg_values: DefaultDict[int, list[str]] = collections.defaultdict(lambda: [])
+    name_values = collections.defaultdict(lambda: [])
     for reg in pwndbg.gdblib.regs.common:
-        reg_values[pwndbg.gdblib.regs[reg]].append(reg)
+        name_values[pwndbg.gdblib.regs[reg]].append(reg)
```
Comment (disconnect3d) [inline pwndbg/commands/telescope.py]: its prboably better to cache `regs = pwndbg.gdblib.regs` above and use here `regs[reg]`

File: pwndbg/commands/telescope.py [hunk 4]
```diff
@@ -167,22 +168,22 @@ def telescope(
         stop = address - ptrsize
         step = -1 * ptrsize
 
-    # Find all registers which show up in the trace, map address to regs
-    regs: dict[int, str] = {}
+    # Find all names which show up in the trace
+    names = {}
     for i in range(start, stop, step):
-        values = list(reg_values[i])
+        values = list(name_values[i])
```
Comment (disconnect3d) [inline pwndbg/commands/telescope.py:176]: Isn't this a list already?

Comment (dmur1) [reply pwndbg/commands/telescope.py:176]: @disconnect3d i think its name_values = collections.defaultdict(lambda: [])

what do you think?

**Comment by dmur1** [comment]
> @dmur1 RE: non-existent stack frame - run `gdb --quiet /bin/ls` and then:
> 
> ```
> pwndbg> set context-sections ''
> ...
> pwndbg> entry
> ...
> pwndbg> stack
> 00:0000│ r13 rsp ret0 0x7fffffffe370 ◂— 0x1
> 01:0008│ ret1         0x7fffffffe378 —▸ 0x7fffffffe604 ◂— '/usr/bin/ls'
> 02:0010│ ret2         0x7fffffffe380 ◂— 0x0
> 03:0018│ rcx          0x7fffffffe388 —▸ 0x7fffffffe610 ◂— 'SHELL=/bin/bash'
> 04:0020│              0x7fffffffe390 —▸ 0x7fffffffe620 ◂— 'PWD=/root/pwndbg'
> 05:0028│              0x7fffffffe398 —▸ 0x7fffffffe631 ◂— 'NIX_PROFILES=/nix/var/nix/profiles/default /root/.nix-profile'
> 06:0030│              0x7fffffffe3a0 —▸ 0x7fffffffe66f ◂— 'LOGNAME=root'
> 07:0038│              0x7fffffffe3a8 —▸ 0x7fffffffe67c ◂— 'XDG_SESSION_TYPE=tty'
> 
> pwndbg> set $rsp=0x1234
> 
> pwndbg> stack
> Exception occurred: telescope: 'NoneType' object has no attribute 'start' (<class 'AttributeError'>)
> For more info invoke `set exception-verbose on` and rerun the command
> or debug it by yourself with `set exception-debugger on`
> pwndbg> set exception-verbose on
> Set whether to print a full stacktrace for exceptions raised in Pwndbg commands to 'on'.
> pwndbg> stack
> ╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
> │ /root/pwndbg/pwndbg/commands/__init__.py:190 in __call__                                         │
> │                                                                                                  │
> │   187 │                                                                                          │
> │   188 │   def __call__(self, *args: Any, **kwargs: Any) -> str | None:                           │
> │   189 │   │   try:                                                                               │
> │ ❱ 190 │   │   │   return self.function(*args, **kwargs)                                          │
> │   191 │   │   except TypeError as te:                                                            │
> │   192 │   │   │   print(f"{self.function.__name__.strip()!r}: {self.function.__doc__.strip()}"   │
> │   193 │   │   │   pwndbg.exception.handle(self.function.__name__)                                │
> │                                                                                                  │
> │ /root/pwndbg/pwndbg/commands/__init__.py:346 in _OnlyWhenRunning                                 │
> │                                                                                                  │
> │   343 │   @functools.wraps(function)                                                             │
> │   344 │   def _OnlyWhenRunning(*a: Any, **kw: Any) -> Optional[T]:                               │
> │   345 │   │   if pwndbg.gdblib.proc.alive:                                                       │
> │ ❱ 346 │   │   │   return function(*a, **kw)                                                      │
> │   347 │   │   else:                                                                              │
> │   348 │   │   │   print(f"{function.__name__}: The program is not being run.")                   │
> │   349 │   │   │   return None                                                                    │
> │                                                                                                  │
> │ /root/pwndbg/pwndbg/commands/telescope.py:159 in telescope                                       │
> │                                                                                                  │
> │   156 │   for reg in pwndbg.gdblib.regs.common:                                                  │
> │   157 │   │   name_values[pwndbg.gdblib.regs[reg]].append(reg)                                   │
> │   158 │                                                                                          │
> │ ❱ 159 │   for idx, retaddr in enumerate(pwndbg.gdblib.stack.yield_return_addresses()):           │
> │   160 │   │   name_values[retaddr].append(f"ret{idx}")                                           │
> │   161 │                                                                                          │
> │   162 │   if not inverse:                                                                        │
> │                                                                                                  │
> │ /root/pwndbg/pwndbg/gdblib/stack.py:171 in yield_return_addresses                                │
> │                                                                                                  │
> │   168 │   │   frame = frame.older()                                                              │
> │   169 │                                                                                          │
> │   170 │   # Find all of them on the stack                                                        │
> │ ❱ 171 │   start, stop = stack.start, stack.end                                                   │
> │   172 │   while addresses and start < sp < stop:                                                 │
> │   173 │   │   value = pwndbg.gdblib.memory.u(sp)                                                 │
> │   174                                                                                            │
> ╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
> AttributeError: 'NoneType' object has no attribute 'start'
> ```
> 
> Pls fix :)

fixed :)

File: pwndbg/commands/telescope.py (hunk 4 above)
Comment (OBarronCS) [reply pwndbg/commands/telescope.py:176]: It's already a list in this case (`print(type(name_values[i]))` to show this). In line 180, we also iterate name_values[i+width] as a list.

File: pwndbg/commands/telescope.py [hunk 5]
```diff
@@ -153,10 +154,12 @@ def telescope(
         count -= address
         count = max(math.ceil(count / ptrsize), 1)
 
-    # Map of address to register string
-    reg_values: DefaultDict[int, list[str]] = collections.defaultdict(lambda: [])
-    for reg in pwndbg.gdblib.regs.common:
-        reg_values[pwndbg.gdblib.regs[reg]].append(reg)
+    name_values = collections.defaultdict(lambda: [])
+    for reg in regs.common:
```
Comment (OBarronCS) [inline pwndbg/commands/telescope.py:158]: Perhaps we add a type + comment here:
```
# Map of address to string (register name, retX)
: DefaultDict[int, list[str]]
```
I had to look ahead and read the rest of the code to see what the types/structure of the data here was,  a type would add some readability/save time for future contributors.

File: pwndbg/commands/telescope.py [hunk 6]
```diff
@@ -167,22 +170,22 @@ def telescope(
         stop = address - ptrsize
         step = -1 * ptrsize
 
-    # Find all registers which show up in the trace, map address to regs
-    regs: dict[int, str] = {}
+    # Find all names which show up in the trace
+    names = {}
```
Comment (OBarronCS) [inline pwndbg/commands/telescope.py:174]: names: dict[int, str] = {}

File: pwndbg/commands/telescope.py [hunk 7]
```diff
@@ -214,32 +217,28 @@ def collapse_repeating_values() -> None:
...
-                idx_offset,
-                delimiter,
-                line_offset,
-                separator,
-            )
-        ) + " ".join(
+
+        line = " ".join(
             (
-                regs_or_frame_offset(addr, bp, regs, longest_regs),
```
Comment (OBarronCS) [inline pwndbg/commands/telescope.py:242]: This is no longer called - frame offset isn't being printed.

File: pwndbg/commands/telescope.py [hunk 8]
```diff
@@ -214,32 +217,28 @@ def collapse_repeating_values() -> None:
...
+                T.offset(
+                    "%02x%s%04x%s"
+                    % (
+                        i + telescope.offset,
+                        delimiter,
+                        addr - start + (telescope.offset * ptrsize),
+                        separator,
+                    )
+                ),
+                T.register(names[addr].ljust(longest_names)),
```
Comment (OBarronCS) [inline pwndbg/commands/telescope.py:241]: There are a couple changes that were discarded during the rebase that we should continue to have:

To bring back both stackframe base offset printing and looks like some logic for reverse printing (#1978), could replace whole section with:

```python
        if inverse:
            line_offset = addr - (stop + ptrsize) + (telescope.offset * ptrsize)
            idx_offset = int((start - stop - ptrsize) / ptrsize) - (i + telescope.offset)
        else:
            line_offset = addr - start + (telescope.offset * ptrsize)
            idx_offset = i + telescope.offset

        line = T.offset(
            "%02x%s%04x%s"
            % (
                idx_offset,
                delimiter,
                line_offset,
                separator,
            )
        ) + " ".join(
            (
                regs_or_frame_offset(addr, bp, names, longest_names),
                pwndbg.chain.format(addr),
            )
        )
```

**Comment by gsingh93** [comment]
@dmur1 could you rebase this?

**Comment by disconnect3d** [comment]
@dmur1 ping about rebase/conflicts

**Comment by disconnect3d** [comment]
This is a bit stale and has lots of conflicts. Closing for now.

**Pull request closed without merging**